rocAL/source/readers/image/numpy_data_reader.cpp

```diff
@@ -0,0 +1,642 @@
+        if (_header_parsing_failed) { \
+            return; \
+        } \
+    } while (0)
+
+NumpyDataReader::NumpyDataReader() {
+    _loop = false;
+    _shuffle = false;
+    _file_count_all_shards = 0;
+}
+
+Reader::Status NumpyDataReader::initialize(ReaderConfig desc) {
+    auto ret = Reader::Status::OK;
+    _folder_path = desc.path();
+    _file_list_path = desc.file_list_path();
+    _shard_id = desc.get_shard_id();
+    _shard_count = desc.get_shard_count();
+    _batch_size = desc.get_batch_size();
+    _shuffle = desc.shuffle();
+    _loop = desc.loop();
+    _meta_data_reader = desc.meta_data_reader();
+    _sharding_info = desc.get_sharding_info();
+    _pad_last_batch_repeated = _sharding_info.pad_last_batch_repeated;
+    _stick_to_shard = _sharding_info.stick_to_shard;
+    _shard_size = _sharding_info.shard_size;
+    _files = desc.get_files();
+    _seed = desc.seed();
+    ret = subfolder_reading();
+    _file_headers.resize(_file_names.size());
+    // shuffle dataset if set
+    if (ret == Reader::Status::OK && _shuffle) {
+        std::mt19937 rng(_seed);
+        std::shuffle(_file_names.begin() + _shard_start_idx_vector[_shard_id],
+                     _file_names.begin() + _shard_end_idx_vector[_shard_id], rng);
+    }
+    return ret;
+}
+
+void NumpyDataReader::incremenet_read_ptr() {
+    _read_counter++;
+    increment_curr_file_idx(_file_names.size());
+}
+
+size_t NumpyDataReader::open() {
+    auto file_path = _file_names[_curr_file_idx];       // Get current file name
+    _curr_file_header = _file_headers[_curr_file_idx];  // Get current file header
+    incremenet_read_ptr();
+    _last_file_path = _last_id = file_path;
+    auto last_slash_idx = _last_id.find_last_of("\\/");
+    if (std::string::npos != last_slash_idx) {
+        _last_id.erase(0, last_slash_idx + 1);
+    }
+
+    _header_parsing_failed = false;
+    auto ret = get_header_from_cache(file_path, _curr_file_header);
+    if (!ret) {
+        parse_header(_curr_file_header, file_path);
+        if(_header_parsing_failed) {
+            ERR("Numpy header parsing failed");
+            return 0;
+        }
+        update_header_cache(file_path, _curr_file_header);
+    } else {
+        _current_file_ptr = std::fopen(file_path.c_str(), "rb");
+        if (_current_file_ptr == nullptr) {
+            ERR("Could not open file " + file_path + ": " + std::strerror(errno));
+            return 0;
+        }
+    }
+    fseek(_current_file_ptr, 0, SEEK_SET);  // Take the file pointer back to the start
+    return _curr_file_header.numpy_data_nbytes();  // Returns the numpy array data size (in bytes)
+}
+
+bool NumpyDataReader::get_header_from_cache(const std::string& file_name, NumpyHeaderData& header) {
+    std::unique_lock<std::mutex> cache_lock(_cache_mutex);
+    auto it = _header_cache.find(file_name);
+    if (it == _header_cache.end()) {
+        return false;
+    } else {
+        header = it->second;
+        return true;
+    }
+}
+
```

Comment:
please change to switch statement

Reply:
Cannot use switch case for get_dtype function since strings are not supported in switch. maps also cannot be used since we need to throw an error for the unsupported types.

Reply:
That is not true. But you can simply handle that with one if
if (any unsupported formats) return NONE;   // add a new type in enum
else return map(numpy to tensor_type)

Reply:
Added a map with supported dtypes and the numpy strings. Checks whether dtype string is present inside the map if not prints an error saying dtype not supported in rocAL and continue to the next sample